components/places/src/db/coop_transaction.rs

```diff
@@ -0,0 +1,199 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::db::PlacesDb;
+use crate::error::*;
+use rusqlite::{Connection, TransactionBehavior};
+use sql_support::{ConnExt, UncheckedTransaction};
+use std::ops::Deref;
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+
+/// This implements "cooperative transactions" for places. It relies on our
+/// decision to have exactly 1 general purpose "writer" connection and exactly
+/// one "sync writer" - ie, exactly 2 write connections.
+///
+/// The idea is that anything that uses the sync connection should use
+/// `time_chunked_transaction`. Code using this should regularly call
+/// `maybe_commit()`, and every second, will commit the transaction and start
+/// a new one.
+///
+/// This means that in theory the other writable connection can start
+/// transactions and should block for a max of 1 second - well under the 5
+/// seconds before that other writer will fail with a SQLITE_BUSY or similar error.
+///
+/// However, in practice we see the writer thread being starved - even though
+/// it's waiting for a lock, the sync thread still manages to re-get the lock.
+/// In other words, the locks used by sqlite aren't "fair".
+///
+/// So we mitigate this with a simple approach that works fine within our
+/// "exactly 2 writers" constraints:
+/// * Each database connection shares a mutex.
+/// * Before starting a transaction, each connection locks the mutex.
+/// * It then starts an "exclusive" transaction - because sqlite now holds a
```

Comment:
I think we want `Immediate`, not `Exclusive`. IIUC:

* `Deferred`, the default, takes a read lock, allowing other readers and writers until its first write. We don't want that, or our first write might fail with `SQLITE_BUSY`.
* `Exclusive` blocks all other writers _and_ readers, which we don't want. (Unless those readers use `PRAGMA read_uncommitted = 1`, which will let them see the uncommitted state of all transactions. 😓)
* `Immediate` takes a write lock, but still allows readers.

Does that sound right to you, am I misunderstanding?

Reply:
I remember one of the difference between our usage and the unlock_notify example was the use of Immediate vs Exclusive. I'm assuming this doesn't matter, but just checking.

Reply:
Last week when playing with this I could reproduce the existing problem when using `Immediate` locks, but not `Exclusive`, which I thought was odd but forgot to come back to.

Fortunately though, that example does now work fine with `Immediate`, so yay :)